Testnet/Airdrop/Airdrop/Airdrop.cs

```diff
@@ -0,0 +1,202 @@
+    public ulong AmountToDistribute
+    {
+        get
+        {
+            ulong amount = PersistentState.GetUInt64(nameof(AmountToDistribute));
+            if (amount == 0 && RegistrationIsClosed)
+            {
+                amount = TotalSupply / NumberOfRegistrants;
+                AmountToDistribute = amount;
+            }
+
+            return amount;
+        }
+
+        private set => PersistentState.SetUInt64(nameof(AmountToDistribute), value);
+    }
+
+    /// <summary>Returns whether or not the registration period is closed.</summary>
+    public bool RegistrationIsClosed
+    {
+        get
+        {
+            bool isClosed = PersistentState.GetBool(nameof(RegistrationIsClosed));
+            if (!isClosed && EndBlock > 0)
+            {
+                isClosed = Block.Number > EndBlock;
+                RegistrationIsClosed = isClosed;
+            }
+
+            return isClosed;
+        }
+
+        private set => PersistentState.SetBool(nameof(RegistrationIsClosed), value);
+    }
+
+    /// <summary>Returns the status of any given address.</summary>
+    public Status GetAccountStatus(Address address)
+    {
+        return PersistentState.GetStruct<Status>($"Status:{address}");
+    }
+
+    /// <summary>Sets the status for a given address.</summary>
+    private void SetAccountStatus(Address address, Status status)
+    {
+        PersistentState.SetStruct($"Status:{address}", status);
+    }
+
+    /// <summary>Validates and registers accounts. See <see cref="AddRegistrantExecute(Address)"/></summary>
+    public bool Register()
+    {
+        return AddRegistrantExecute(Message.Sender);
+    }
+
+    /// <summary>Allows owner of the contract to manually add a new registrant. See <see cref="AddRegistrantExecute(Address)"/></summary>
+    public bool AddRegistrant(Address registrant)
+    {
+        return Message.Sender == Owner && AddRegistrantExecute(registrant);
+    }
+
+    /// <summary>Allows owner to close registration for the airdrop at any time.</summary>
+    public bool CloseRegistration()
+    {
+        if (Message.Sender != Owner)
+        {
+            return false;
+        }
+
+        RegistrationIsClosed = true;
+
+        return true;
+    }
+
+    /// <summary>
+    /// Withdraw funds after registration period has closed. Validates account status and calls the tokens contract
+    /// address that is being airdropped to transfer amount to sender. On success, set senders new status and log it.
+    /// </summary>
+    public bool Withdraw()
+    {
+        bool invalidAccountStatus = GetAccountStatus(Message.Sender) != Status.ENROLLED;
+        if (invalidAccountStatus || !RegistrationIsClosed || AmountToDistribute == 0)
+        {
+            return false;
+        }
+
+        var transferParams = new object[] { Message.Sender, AmountToDistribute };
+
+        ITransferResult result = Call(TokenContractAddress, AmountToDistribute, "TransferTo", transferParams);
```

Comment:
Noting the earlier discussion about choice of `TransferTo` vs. `TransferFrom` - the intended usage of `TransferFrom` is exactly for scenarios like this, where a third party (or contract) needs to transfer tokens on another party's behalf. 

In short, `TransferTo` allows _you_ to transfer tokens that you own, while `TransferFrom` allows _someone else_ to transfer tokens that you own.

For a reason why this might be beneficial, consider the current scenario where `TransferTo` is used and some amount of tokens are assigned to the contract. For some reason, the airdrop is cancelled. Or, maybe there is a bug in the contract, and the tokens cannot be transferred to recipients. The Airdrop contract has no method to return those tokens to the owner, so they are effectively locked up and must either be distributed or burnt. If `TransferFrom` were used, the holder of the airdrop supply could just remove the approval for the old contract, and that would be it.

This affords the party running the airdrop a considerable amount of control, which could be a good or bad thing depending on the circumstances.

Reply:
Not suggesting one approach over the other. Both are valid, depending on the use case, but it's important to understand what happens if something does go wrong. Case in point: the rounding in `AmountToTransfer` could cause at least 1 token to be permanently burnt.